integration-tests/tests/src/tests/notifications.ts

```diff
@@ -0,0 +1,463 @@
+
+import { expect } from "chai";
+import Realm from "realm";
+import { openRealmBeforeEach } from "../hooks";
+
+class ListObject extends Realm.Object {
+  list!: Realm.List<TestObject>;
+  static schema = {
+    name: "ListObject",
+    properties: {
+      list: { type: "list", objectType: "TestObject" },
+    },
+  };
+}
+
+class TestObject extends Realm.Object {
+  doubleCol!: Realm.Types.Double;
+  static schema = {
+    name: "TestObject",
+    properties: {
+      doubleCol: "double",
+    },
+  };
+}
+
+describe("Notifications", () => {
+  let runCount = 0;
+
+  beforeEach(() => (runCount = 0));
+
+  describe("Realm notifications", () => {
+    openRealmBeforeEach({ schema: [TestObject] });
+
+    it("should fire on change", function (this: RealmContext, done) {
+      this.realm.addListener("change", (sender, event) => {
+        expect(sender).deep.equals(this.realm);
+        expect(event).equals("change");
+
+        const objects = this.realm.objects(TestObject);
+
+        expect(objects.length).equals(1);
+        expect(objects[0].doubleCol).equals(42);
+        done();
+      });
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 42 });
+      });
+    });
+
+    it("should fire on multiple changes", function (this: RealmContext, done) {
+      this.realm.addListener("change", (sender, event) => {
+        expect(sender).deep.equals(this.realm);
+        expect(event).equals("change");
+
+        const objects = this.realm.objects(TestObject);
+        runCount++;
+        if (runCount === 1) {
+          expect(objects.length).equals(1);
+          expect(objects[0].doubleCol).equals(1);
+        } else if (runCount == 2) {
+          expect(objects.length).equals(3);
+          expect(objects[1].doubleCol).equals(2);
+          expect(objects[2].doubleCol).equals(3);
+          done();
+        }
+      });
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 1 });
+      });
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 2 });
+        this.realm.create("TestObject", { doubleCol: 3 });
+      });
+    });
+
+    it("implements removeListener", function (this: RealmContext) {
+      const handler = () => {
+        const objects = this.realm.objects(TestObject);
+        runCount++;
+        if (runCount === 1) {
+          expect(objects.length).equals(1);
+          expect(objects[0].doubleCol).equals(1);
+        } else {
+          throw new Error("Should not run after removeListener");
+        }
+      };
+
+      this.realm.addListener("change", handler);
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 1 });
+      });
+
+      this.realm.removeListener("change", handler);
+
+      this.realm.write(() => {
+        this.realm.create("TestObject", { doubleCol: 2 });
+      });
+
+      expect(runCount).equal(1);
+    });
```

Comment:
yeah there is some async weirdness that can happen here, the "proper" test could be to do some kind of timed out done? i.e. give 1 second (or less, like 500ms) to make sure the listener does not get called and consider that good enough?

Reply:
I would rather await something async than relying on timeouts, as they tend to add up when running the entire test suite.

Reply:
> Mocha errors if done is called multiple times.

While this is true, I wouldn't rely on this, because it messes up the report. First Mocha thinks the test pass but then it fails the test. It's ultimately a race between the second call to `done` and the test command exiting green, which could lead to unstable tests.